man/cafe-display-properties-install-systemwide.1
.\" Copyright (C) 2014 Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
.\"
.\" This is free software; you may redistribute it and/or modify
.\" it under the terms of the GNU General Public License as
.\" published by the Free Software Foundation; either version 2,
.\" or (at your option) any later version.
.\"
.\" This is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\"You should have received a copy of the GNU General Public License along
.\"with this program; if not, write to the Free Software Foundation, Inc.,
.\"51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

.TH cafe-display-properties-install-systemwide 1 "2014\-05\-02" "CAFE"
.\"
.SH "NAME"
cafe-display-properties-install-systemwide \- customize CAFE display behavor
.SH "SYNOPSIS"
.B cafe-display-properties-install-systemwide
\fI<SOURCE_FILE> <DEST_NAME>\fR
.SH "DESCRIPTION"
This executable is part of the package 'cafe\-control\-center': The CAFE Control Center.
.PP
This command line tool installs a RANDR profile for multi-monitor setups into
a systemwide location.  The resulting profile will get used when
the RANDR plug-in gets run in \fBcafe-settings-daemon\fR.
.TP
\fI<SOURCE_FILE>\fR
a full pathname, typically /home/username/.config/monitors.xml
.TP
\fI<DEST_NAME>\fR
relative name for the installed file.  This will get put in
the systemwide directory for RANDR configurations,
so the result will typically be /etc/cafe-settings-daemon/xrandr/DEST_NAME
.SH AUTHORS
This manual page was writtenby Mike Gabriel <mike.gabriel@das-netzwerkteam.de>
for the Debian project (but may be used by others).
.SH "SEE ALSO"
.PP
Latest version of the CAFE Desktop User Guide for your platform.
.PP
.BR "cafe-control-center" (1),
.BR "cafe-display-properties" (1),
.BR "cafe-settings-daemon" (1),


NEWS
### cafe-control-center 1.24.1

  * Translations update
  * cafe-theme-info: Use of memory after it is freed
  * build: silent build warnings for distcheck
  * use default CAFE_COMPILE_WARNINGS level
  * Port libslab_handle_g_error to the built-in GLib logging framework
  * cafe-about-me: Fix memory leak
  * appearance: Fix memory leak
  * Do not collect the translation for Icon

### cafe-control-center 1.24.0

  * Translations update
  * cafe-theme-info: fix memory leak
  * cafe-wp-info: fix memory leak
  * glib-genmarshal: Using --header and --body at the same time is deprecated
  * build: show compiler/flags info on configure summary
  * build: do not override WARN_CFLAGS
  * build: use warning flags
  * time-admin: make portable backward file path
  * appearance: backgrounds dir should be portable
  * appearance: remove I/O error on window close
  * autoconf: AC_PROG_CC was called before AX_CHECK_ENABLE_DEBUG
  * Fix build using gcc 10 -fno-common flag

### cafe-control-center 1.23.3

  * Translations update
  * Update copyright to 2020
  * about-me: Fix UI validation
  * mouse: Fix UI validation
  * time-admin: Add license info in time-admin.ui file
  * time-admin: Fix UI validation
  * at-properties: Fix UI validation
  * Search system themes in system data dirs
  * display: Fix UI validation
  * display: Enable help button
  * windows: Add window-properties.ui
  * avoid redundant redeclarations
  * keybindings: Fix schema creation in makefile
  * appearance: Fix desktop file creation
  * shell: Fix typo in keyword name
  * migrate from intltool to gettext
  * default-applications: Load icons as cairo surfaces
  * icons: Force uniform size on icons (#505)
  * remove warning: "__sun" is not defined, evaluates to 0
  * fv: thumbnailer: add font/collection to the supported mimetypes
  * font-viewer: Add font/ttf and font/otf mime types
  * time-admin: fix memory leak
  * time-admin: Don't share «void QuitApp(TimeAdmin *ta)» method
  * default-applications: Remove messenger options from ui file
  * time-admin: Remove unreachable code
  * time-admin: Simplify date & time displaying
  * time-admin: Save button cannot be enabled if NTP sync is active
  * time-admin: Display current time in time-admin.ui on startup
  * time-admin: Rename adjustament to adjustment in time-admin.ui
  * time-admin: A11y time-admin.ui
  * time-admin: Add time-admin.ui
  * default-applications: Fix Word Processor top margin

### cafe-control-center 1.23.2

  * Translations update
  * travis: update cafe-desktop and cafe-settings-daemon required versions
  * mouse: Add support for Acceleration Profiles
  * appearance: fix memory leak
  * appearance: add new font for user
  * Stop propagate error
  * at-properties: migrate from dbus-glib to gdbus

### cafe-control-center 1.23.1

  * Translations update
  * fix build error when not installed gla11y
  * remove lightdm-settings support
  * launch lightdm-settings if installed
  * use current running DM preferences program
  * support lightdm-gtk-greeter-settings as login preferences program
  * replace Tab with space
  * drop libXxf86 dependency
  * Use python3 for gla11y
  * Add scroll tabs with mouse wheel in severals capplets
  * libslab: bookmark-agent.c avoid deprecated g_type_class_add_private
  * font-viewer: avoid deprecated 'g_type_class_add_private'
  * Remove compilation warnings
  * time-admin: Don't set window title for message-box dialog
  * time-admin: Don't set font size/color for regular controls
  * use theme's icon preferences-system-time
  * time-admin: Fix license notice in time-zone.h|c, time-map.h|c
  * capplets: Remove the macro CAFELOCALEDIR where it's not required
  * capplets: Use capplet_init
  * time-admin: shortcuts, images & shared code for buttons
  * common: Remove capplet-stock-icons
  * time-admin: Remove po-timezones
  * Remove trailing whitespaces

### cafe-control-center 1.23.0

  * Translations update
  * bump require of m-s-d and m-d
  * Add orca to Visual combo box if available
  * Add support for checking .ui accessibility with gla11y
  * Do not reload NTP and modify the button arrangement when starting the service
  * Authentication to add modification time
  * Use default color to set label and remove the maximum button
  * use Makefile pkgdatadir
  * g_date_time_unref datetime
  * Display the correct time when selecting the time zone
  * Memory leak: g_error_free(error);
  * Modify cafe-time-admin.desktop.in file format
  * enable to build time-admin for cafe-control-center
  * add time-admin source code
  * add timezone translation files for time-admin
  * add image resources for time-admin applet
  * Populate initial messenger list with irc scheme handler apps
  * Set default icq/irc/ircs/sip/xmpp scheme handler
  * Add instant messenger keybinding into cafe-keybinding-properties
  * Change the icon for instant messaging application
  * default-applications: Add a widget for a default messenger application
  * appearance.ui (cursor size scale)
  * libslab: avoid deprecated g_type_class_add_private
  * Create FUNDING.yml
  * typing-break: avoid deprecated 'g_type_class_add_private'
  * file-transfer-dialog.c: avoid deprecated 'g_type_class_add_private'
  * Fix the icon sizes in Preferred Applications selector
  * capplets: Use gresources for UI files
  * [ci] Add cppcheck html report
  * Fix grammar mistake
  * Migrate from dbus-glib to GDBus
  * Use system data dirs to locate key bindings
  * control-center: Fix memory leak
  * Use GtkSeparatorMenuItem in Apparence preview
  * Remove dead store detected by clang
  * default-applications: Improve UI accessibility
  * Use keyboard device icon like in other hardware devices
  * Use icons provided by themes and remove not used icons
  * help: update copyright
  * [ci] Enable Clang Static Analyzer
  * add installation instructions

### cafe-control-center 1.22.0

  * Translations update
  * Initialize travis support

capplets/display/cafe-display-properties-install-systemwide.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
 *
 * cafe-display-properties-install-systemwide - Install a RANDR profile for the whole system
 *
 * Copyright (C) 2010 Novell, Inc.
 *
 * Authors: Federico Mena Quintero <federico@novell.com>
 *
 * Licensed under the GNU General Public License Version 2
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include "config.h"

#include <errno.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <fcntl.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <glib/gi18n.h>
#include <glib.h>

#define SYSTEM_RANDR_PATH "/etc/cafe-settings-daemon/xrandr"

static void
usage (const char *program_name)
{
	g_print (_("Usage: %s SOURCE_FILE DEST_NAME\n"
		   "\n"
		   "This program installs a RANDR profile for multi-monitor setups into\n"
		   "a systemwide location.  The resulting profile will get used when\n"
		   "the RANDR plug-in gets run in cafe-settings-daemon.\n"
		   "\n"
		   "SOURCE_FILE - a full pathname, typically /home/username/.config/monitors.xml\n"
		   "\n"
		   "DEST_NAME - relative name for the installed file.  This will get put in\n"
		   "            the systemwide directory for RANDR configurations,\n"
		   "            so the result will typically be %s/DEST_NAME\n"),
		 program_name,
		 SYSTEM_RANDR_PATH);
}

static gboolean
is_basename (const char *filename)
{
	if (*filename == '\0')
		return FALSE; /* no empty strings, please */

	for (; *filename; filename++)
		if (G_IS_DIR_SEPARATOR (*filename))
			return FALSE;

	return TRUE;
}

static gboolean
copy_file (int source_fd, int dest_fd)
{
	char buf[1024];
	int num_read;
	int num_written;

	while (TRUE) {
		char *p;

		num_read = read (source_fd, buf, sizeof (buf));
		if (num_read == 0)
			break;

		if (num_read == -1) {
			if (errno == EINTR)
				continue;
			else
				return FALSE;
		}

		p = buf;
		while (num_read > 0) {
			num_written = write (dest_fd, p, num_read);
			if (num_written == -1) {
				if (errno == EINTR)
					continue;
				else
					return FALSE;
			}

			num_read -= num_written;
			p += num_written;
		}
	}

	return TRUE;
}

/* This is essentially a "please copy a file to a privileged location" program.
 * We try to be paranoid in the following ways:
 *
 * - We copy only regular files, owned by the user who called pkexec(1), to
 *   avoid attacks like "copy a file that I'm not allowed to read into a
 *   world-readable location".
 *
 * - We copy only to a well-known directory.
 *
 * - We try to avoid race conditions.  We only fstat() files that we have open
 *   to avoid files moving under our feet.  We only create files in directories
 *   that we have open.
 *
 * - We replace the destination file atomically.
 */

int
main (int argc, char **argv)
{
	uid_t uid, euid;
	const char *source_filename;
	const char *dest_name;
	const char *pkexec_uid_str;
	int pkexec_uid;
	struct stat statbuf;
	int err;
	int source_fd;
	int dest_fd;
	char template[100];

	setlocale (LC_ALL, "");
	bindtextdomain (GETTEXT_PACKAGE, CAFELOCALEDIR);
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);

	/* We only run as root */
	uid = getuid ();
	euid = geteuid ();
	if (uid != 0 || euid != 0) {
		/* Translators: only able to install RANDR profiles as root */
		g_print ("%s\n", _("This program can only be used by the root user"));
		return EXIT_FAILURE;
	}

	/* Usage: gsd-xrandr-install-systemwide SOURCE_FILE DEST_NAME */

	if (argc != 3) {
		usage (argv[0]);
		return EXIT_FAILURE;
	}

	source_filename = argv[1];
	dest_name = argv[2];

	/* Absolute source filenames only, please */

	if (!g_path_is_absolute (source_filename)) {
		g_print ("%s\n", _("The source filename must be absolute"));
		return EXIT_FAILURE;
	}

	/* We only copy regular files */

	source_fd = open (source_filename, O_RDONLY);
	if (source_fd == -1) {
		err = errno;

		/* Translators: first %s is a filename; second %s is an error message */
		g_print (_("Could not open %s: %s\n"),
			 source_filename,
			 g_strerror (err));
		return EXIT_FAILURE;
	}

	if (fstat (source_fd, &statbuf) != 0) {
		err = errno;
		/* Translators: first %s is a filename; second %s is an error message */
		g_print (_("Could not get information for %s: %s\n"),
			 source_filename,
			 g_strerror (err));
		return EXIT_FAILURE;
	}

	if (!S_ISREG (statbuf.st_mode)) {
		g_print (_("%s must be a regular file\n"),
			 source_filename);
		return EXIT_FAILURE;
	}

	/* We only copy files that are really owned by the calling user */

	pkexec_uid_str = g_getenv ("PKEXEC_UID");
	if (pkexec_uid_str == NULL) {
		g_print ("%s\n", _("This program must only be run through pkexec(1)"));
		return EXIT_FAILURE;
	}

	if (sscanf (pkexec_uid_str, "%d", &pkexec_uid) != 1) {
		g_print ("%s\n", _("PKEXEC_UID must be set to an integer value"));
		return EXIT_FAILURE;
	}

	if (statbuf.st_uid != pkexec_uid) {
		/* Translators: we are complaining that a file must be really owned by the user who called this program */
		g_print (_("%s must be owned by you\n"), source_filename);
		return EXIT_FAILURE;
	}

	/* We only accept basenames for the destination */

	if (!is_basename (dest_name)) {
		/* Translators: here we are saying that a plain filename must look like "filename", not like "some_dir/filename" */
		g_print (_("%s must not have any directory components\n"),
			 dest_name);
		return EXIT_FAILURE;
	}

	/* Chdir to the destination directory to keep it open... */

	if (chdir (SYSTEM_RANDR_PATH) != 0) {
		g_print (_("%s must be a directory\n"), SYSTEM_RANDR_PATH);
		return EXIT_FAILURE;
	}

	/* ... and open our temporary destination file right there */

	strcpy (template, "gsd-XXXXXX");
	dest_fd = g_mkstemp_full (template, O_WRONLY, 0644);
	if (dest_fd == -1) {
		err = errno;
		/* Translators: the first %s/%s is a directory/filename; the last %s is an error message */
		g_print (_("Could not open %s/%s: %s\n"),
			 SYSTEM_RANDR_PATH,
			 template,
			 g_strerror (err));
		return EXIT_FAILURE;
	}

	/* Do the copy */

	if (!copy_file (source_fd, dest_fd)) {
		/* If something went wrong, remove the destination file to avoid leaving trash around */
		unlink (template);
		return EXIT_FAILURE;
	}

	/* Rename to the final filename */

	if (rename (template, dest_name) != 0) {
		err = errno;
		unlink (template);
		g_print (_("Could not rename %s to %s: %s\n"),
			 template,
			 dest_name,
			 g_strerror (err));
		return EXIT_FAILURE;
	}

	/* Whew!  We'll leave the final closing of the files to the almighty kernel. */

	return EXIT_SUCCESS;
}
